File: E2ETestResults/CoreSubset/CoreSlim/Utility/WindowHelper.cs
// ---------------------------------------------------------------------------------------------------------------------
// <copyright file="WindowHelper.cs" company="Tableau Software">
//   This file is the copyrighted property of Tableau Software and is protected by registered patents and other
//   applicable U.S. and international laws and regulations.
//
//   Unlicensed use of the contents of this file is prohibited. Please refer to the NOTICES.txt file for further details.
// </copyright>
// ---------------------------------------------------------------------------------------------------------------------

namespace Tableau.JavaScript.CoreSlim
{
    using System;
    using System.Diagnostics.CodeAnalysis;
    using System.Html;
    using System.Html.Editing;
    using System.Runtime.CompilerServices;

    /// <summary>
    /// Contains a set of helper functions for a <see cref="WindowInstance"/>.
    /// </summary>
    public class WindowHelper
    {
        //// ===========================================================================================================
        //// Member Variables
        //// ===========================================================================================================

        private static readonly Func<WindowInstance, int> InnerWidthFunc;

        private static readonly Func<WindowInstance, int> InnerHeightFunc;

        private static readonly Func<WindowInstance, int> ClientWidthFunc;

        private static readonly Func<WindowInstance, int> ClientHeightFunc;

        private static readonly Func<WindowInstance, int> PageXOffsetFunc;

        private static readonly Func<WindowInstance, int> PageYOffsetFunc;

        private static readonly Func<WindowInstance, int> ScreenLeftFunc;

        private static readonly Func<WindowInstance, int> ScreenTopFunc;

        private static readonly Func<WindowInstance, int> OuterWidthFunc;

        private static readonly Func<WindowInstance, int> OuterHeightFunc;

        private static Func<Action, int> requestAnimationFrameFunc;
        private static Action<int> cancelAnimationFrameFunc;

        private readonly WindowInstance window;

        static WindowHelper()
        {
            // polyfills for various window properties - http://www.quirksmode.org/mobile/viewports2.html

            // InnerWidth
            if (Script.In(typeof(Window), "innerWidth"))
            {
                InnerWidthFunc = delegate(WindowInstance w) { return w.InnerWidth; };
            }
            else
            {
                InnerWidthFunc = delegate(WindowInstance w) { return Script.Reinterpret<HtmlElement>(w.Document.DocumentElement).OffsetWidth; };
            }

            // OuterWidth
            if (Script.In(typeof(Window), "outerWidth"))
            {
                OuterWidthFunc = delegate(WindowInstance w) { return w.OuterWidth; };
            }
            else
            {
                // TODO: This is not correct, but it'll do for now. B147312
                OuterWidthFunc = InnerWidthFunc;
            }

            // InnerHeight
            if (Script.In(typeof(Window), "innerHeight"))
            {
                InnerHeightFunc = delegate(WindowInstance w) { return w.InnerHeight; };
            }
            else
            {
                InnerHeightFunc = delegate(WindowInstance w) { return Script.Reinterpret<HtmlElement>(w.Document.DocumentElement).OffsetHeight; };
            }

            // OuterHeight
            if (Script.In(typeof(Window), "outerHeight"))
            {
                OuterHeightFunc = delegate(WindowInstance w) { return w.OuterHeight; };
            }
            else
            {
                // TODO: This is not correct, but it'll do for now. B147312
                OuterHeightFunc = InnerHeightFunc;
            }

            // ClientWidth
            if (Script.In(typeof(Window), "clientWidth"))
            {
                ClientWidthFunc = delegate(WindowInstance w) { return TypeUtil.GetField<int>(w, "clientWidth"); };
            }
            else
            {
                ClientWidthFunc = delegate(WindowInstance w) { return w.Document.DocumentElement.ClientWidth; };
            }

            // ClientHeight
            if (Script.In(typeof(Window), "clientHeight"))
            {
                ClientHeightFunc = delegate(WindowInstance w) { return TypeUtil.GetField<int>(w, "clientHeight"); };
            }
            else
            {
                ClientHeightFunc = delegate(WindowInstance w) { return w.Document.DocumentElement.ClientHeight; };
            }

            // PageXOffsetFunc
            if (Script.IsValue(Window.Self.PageXOffset))
            {
                PageXOffsetFunc = delegate(WindowInstance w) { return w.PageXOffset; };
            }
            else
            {
                PageXOffsetFunc = delegate(WindowInstance w) { return w.Document.DocumentElement.ScrollLeft; };
            }

            // PageYOffsetFunc
            if (Script.IsValue(Window.Self.PageYOffset))
            {
                PageYOffsetFunc = delegate(WindowInstance w) { return w.PageYOffset; };
            }
            else
            {
                PageYOffsetFunc = delegate(WindowInstance w) { return w.Document.DocumentElement.ScrollTop; };
            }

            // ScreenLeftFunc
            if (Script.In(typeof(Window), "screenLeft"))
            {
                ScreenLeftFunc = delegate(WindowInstance w) { return ((dynamic)w).screenLeft; };
            }
            else
            {
                ScreenLeftFunc = delegate(WindowInstance w) { return w.ScreenX; };
            }

            // ScreenTopFunc
            if (Script.In(typeof(Window), "screenTop"))
            {
                ScreenTopFunc = delegate(WindowInstance w) { return ((dynamic)w).screenTop; };
            }
            else
            {
                ScreenTopFunc = delegate(WindowInstance w) { return w.ScreenY; };
            }

            // Polyfill for RequestAnimationFrame - http://paulirish.com/2011/requestanimationframe-for-smart-animating/
            {
                const string DefaultRequestName = "requestAnimationFrame";
                const string DefaultCancelName = "cancelAnimationFrame";
                string[] vendors = new string[] { "ms", "moz", "webkit", "o" };
                string requestFuncName = null;
                string cancelFuncName = null;

                if (Script.In(typeof(Window), DefaultRequestName))
                {
                    requestFuncName = DefaultRequestName;
                }

                if (Script.In(typeof(Window), DefaultCancelName))
                {
                    cancelFuncName = DefaultCancelName;
                }

                for (int ii = 0; ii < vendors.Length && (requestFuncName == null || cancelFuncName == null); ++ii)
                {
                    string vendor = vendors[ii];
                    string funcName = vendor + "RequestAnimationFrame";
                    if (requestFuncName == null && Script.In(typeof(Window), funcName))
                    {
                        requestFuncName = funcName;
                    }

                    if (cancelFuncName == null)
                    {
                        //string cancelFuncName = (Func<Action<int>, int>)Script.Literal("window[{0}]", requestFuncName);
                        funcName = vendor + "CancelAnimationFrame";
                        if (Script.In(typeof(Window), funcName))
                        {
                            cancelFuncName = funcName;
                        }
                        funcName = vendor + "CancelRequestAnimationFrame";
                        if (Script.In(typeof(Window), funcName))
                        {
                            cancelFuncName = funcName;
                        }
                    }
                }

                if (requestFuncName != null)
                {
                    requestAnimationFrameFunc = delegate(Action callback)
                    {
                        return TypeUtil.GetField<Func<Action, int>>(typeof(Window), requestFuncName)(callback);
                    };
                }
                else
                {
                    SetDefaultRequestAnimationFrameImpl();
                }

                if (cancelFuncName != null)
                {
                    cancelAnimationFrameFunc = delegate(int animationId)
                    {
                        TypeUtil.GetField<Action<int>>(typeof(Window), cancelFuncName)(animationId);
                    };
                }
                else
                {
                    cancelAnimationFrameFunc = delegate(int id) { Window.ClearTimeout(id); };
                }
            }
        }

        //// ===========================================================================================================
        //// Constructors
        //// ===========================================================================================================

        public WindowHelper(WindowInstance window)
        {
            this.window = window;
        }

        //// ===========================================================================================================
        //// Properties
        //// ===========================================================================================================

        /// <summary>
        /// Gets the window pageXOffset, or equivalent.
        /// </summary>
        public int PageXOffset
        {
            get
            {
                return PageXOffsetFunc(this.window);
            }
        }

        /// <summary>
        /// Gets the window pageYOffset, or equivalent.
        /// </summary>
        public int PageYOffset
        {
            get
            {
                return PageYOffsetFunc(this.window);
            }
        }

        /// <summary>
        /// Gets the window pageXOffset, or equivalent.
        /// </summary>
        public int ClientWidth
        {
            get
            {
                return ClientWidthFunc(this.window);
            }
        }

        /// <summary>
        /// Gets the window pageYOffset, or equivalent.
        /// </summary>
        public int ClientHeight
        {
            get
            {
                return ClientHeightFunc(this.window);
            }
        }

        /// <summary>
        /// Gets the window innerWidth, or equivalent.
        /// </summary>
        public int InnerWidth
        {
            get
            {
                return InnerWidthFunc(this.window);
            }
        }

        /// <summary>
        /// Gets the window outerWidth, or innerWidth if unavailable
        /// </summary>
        public int OuterWidth
        {
            get
            {
                return OuterWidthFunc(this.window);
            }
        }

        /// <summary>
        /// Gets the window innerHeight, or equivalent.
        /// </summary>
        public int InnerHeight
        {
            get
            {
                return InnerHeightFunc(this.window);
            }
        }

        /// <summary>
        /// Gets the window outerHeight, or innerHeight if unavailable.
        /// </summary>
        public int OuterHeight
        {
            get
            {
                return OuterHeightFunc(this.window);
            }
        }

        /// <summary>
        /// Gets the window screenLeft, or equivalent.
        /// </summary>
        public int ScreenLeft
        {
            get
            {
                return ScreenLeftFunc(this.window);
            }
        }

        /// <summary>
        /// Gets the window screenTop, or equivalent
        /// </summary>
        public int ScreenTop
        {
            get
            {
                return ScreenTopFunc(this.window);
            }
        }

        /// <summary>
        /// Gets the window's self.
        /// This method only exists to support testing.
        /// </summary>
        public static WindowInstance WindowSelf
        {
            get
            {
                return Window.Self;
            }
        }

        /// <summary>
        /// Get the current window or document Selection object
        /// </summary>
        /// <returns>The selection object, or null if none</returns>
        public static Selection Selection
        {
            get
            {
                if (TypeUtil.HasMethod(typeof(Window), "getSelection"))
                {
                    return Window.GetSelection();
                }
                else if (TypeUtil.HasMethod(typeof(Document), "getSelection"))
                {
                    return Document.GetSelection();
                }

                return null;
            }
        }

        //// ===========================================================================================================
        //// Methods
        //// ===========================================================================================================

        /// <summary>
        /// Calls <see cref="Window.Close"/> on the given window.
        /// This method only exists to support testing.
        /// </summary>
        public static void Close(WindowInstance window)
        {
            window.Close();
        }

        /// <summary>
        /// Gets the window opener.
        /// This method only exists to support testing.
        /// </summary>
        public static WindowInstance GetOpener(WindowInstance window)
        {
            return window.Opener;
        }

        /// <summary>
        /// Gets the current window location.
        /// This method only exists to support testing as the standard window.location cannot be mocked safely.
        /// </summary>
        public static Location GetLocation(WindowInstance window)
        {
            return window.Location;
        }

        /// <summary>
        /// Gets the current window location.
        /// This method only exists to support testing as the standard window.location cannot be mocked safely.
        /// </summary>
        public static string GetPathAndSearch(WindowInstance window)
        {
            return window.Location.Pathname + window.Location.Search;
        }

        /// <summary>
        /// Sets the href property on the window's location.
        /// This method only exists to support testing as the standard window.location cannot be mocked safely.
        /// </summary>
        /// <param name="window">The window containing the location</param>
        /// <param name="href">The href to set</param>
        public static void SetLocationHref(WindowInstance window, string href)
        {
            window.Location.Href = href;
        }

        /// <summary>
        /// Calls <see cref="Location.Replace"/> on the given window.
        /// This method only exists to support testing as the standard window.location cannot be mocked safely.
        /// </summary>
        public static void LocationReplace(WindowInstance window, string url)
        {
            window.Location.Replace(url);
        }

        [AlternateSignature]
        [SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters")]
        public static extern WindowInstance Open(string href, string target);

        /// <summary>
        /// Calls <see cref="Window.Open(string,string,string)"/>.
        /// This method only exists to support testing as the standard window.location cannot be mocked safely.
        /// </summary>
        /// <param name="href"></param>
        /// <param name="target"></param>
        /// <param name="options"></param>
        public static WindowInstance Open(string href, string target, string options)
        {
            return Window.Open(href, target, options);
        }

        /// <summary>
        /// Calls window.location.reload.
        /// This method only exists to support testing as the standard window.location cannot be mocked safely.
        /// </summary>
        public static void Reload(WindowInstance w, bool forceGet = false)
        {
            w.Location.Reload(forceGet);
        }

        /// <summary>
        /// Requests an animation frame
        /// Falls back to using Window.SetTimeout if the browser does not support RequestAnimationFrame
        /// </summary>
        /// <param name="action">Action to execute</param>
        /// <returns>animation id used to cancel the animation</returns>
        public static int RequestAnimationFrame(Action action)
        {
            return requestAnimationFrameFunc(action);
        }

        /// <summary>
        /// Cancels an animation
        /// </summary>
        /// <param name="animationId">id of animation to cancel</param>
        public static void CancelAnimationFrame(int animationId)
        {
            if (Script.IsValue(animationId))
            {
                cancelAnimationFrameFunc(animationId);
            }
        }

        /// <summary>
        /// This method exists so that Window.SetTimeout can be mocked
        /// </summary>
        public static void SetTimeout(Action callback, int milliseconds)
        {
            Window.SetTimeout(callback, milliseconds);
        }

        public static void AddListener(EventTarget windowParam, string eventName, HtmlEventHandler messageListener)
        {
            if (Script.In(windowParam, "addEventListener"))
            {
                windowParam.AddEventListener(eventName, messageListener, false);
            }
            else
            {
                ((dynamic)windowParam).attachEvent("on" + eventName, messageListener);
            }
        }

        public static void RemoveListener(EventTarget window, string eventName, HtmlEventHandler messageListener)
        {
            if (Script.In(window, "removeEventListener"))
            {
                window.RemoveEventListener(eventName, messageListener, false);
            }
            else
            {
                ((dynamic)window).detachEvent("on" + eventName, messageListener);
            }
        }

        /// <summary>
        /// Sets the RequestAnimationFrame implementation to the fallback if there is no browser support.
        /// This is in its own method so the closure over 'lastTime' is smaller.
        /// </summary>
        private static void SetDefaultRequestAnimationFrameImpl()
        {
            int lastTime = 0;
            requestAnimationFrameFunc = delegate(Action callback)
            {
                int curTime = (int)(new JsDate().GetTime());
                int timeToCall = Math.Max(0, 16 - (curTime - lastTime));
                lastTime = curTime + timeToCall;
                int id = Window.SetTimeout(callback, timeToCall);
                return id;
            };
        }

        /// <summary>
        /// Clears the selection object at the window/document level
        /// </summary>
        public static void ClearSelection()
        {
            var selection = Selection;

            if (selection != null)
            {
                if (TypeUtil.HasMethod(selection, "removeAllRanges"))
                {
                    selection.RemoveAllRanges();
                }
                else if (TypeUtil.HasMethod(selection, "empty"))
                {
                    Script.InvokeMethod(selection, "empty");
                }
            }
        }
    }
}

File: E2ETestResults/CoreSubset/CoreSlim/Utility/WindowHelper.ts
import 'mscorlib';

import { TypeUtil } from 'NativeJsTypeDefs';

/**
 * Contains a set of helper functions for a {@link WindowInstance}.
 */
export class WindowHelper {
  private static readonly innerWidthFunc: (windowInstance: Object) => number;

  private static readonly innerHeightFunc: (windowInstance: Object) => number;

  private static readonly clientWidthFunc: (windowInstance: Object) => number;

  private static readonly clientHeightFunc: (windowInstance: Object) => number;

  private static readonly pageXOffsetFunc: (windowInstance: Object) => number;

  private static readonly pageYOffsetFunc: (windowInstance: Object) => number;

  private static readonly screenLeftFunc: (windowInstance: Object) => number;

  private static readonly screenTopFunc: (windowInstance: Object) => number;

  private static readonly outerWidthFunc: (windowInstance: Object) => number;

  private static readonly outerHeightFunc: (windowInstance: Object) => number;

  private static requestAnimationFrameFunc: (action: () => void) => number;

  private static cancelAnimationFrameFunc: (int32: number) => void;

  private readonly window: Object;

  // Converted from the C# static constructor - it would be good to convert this
  // block to inline initializations.
  public static __ctor() {
    if ('innerWidth' in window) {
      WindowHelper.innerWidthFunc = (w: Object) => {
        return w.innerWidth;
      };
    } else {
      WindowHelper.innerWidthFunc = (w: Object) => {
        return w.document.documentElement.offsetWidth;
      };
    }
    if ('outerWidth' in window) {
      WindowHelper.outerWidthFunc = (w: Object) => {
        return w.outerWidth;
      };
    } else {
      WindowHelper.outerWidthFunc = WindowHelper.innerWidthFunc;
    }
    if ('innerHeight' in window) {
      WindowHelper.innerHeightFunc = (w: Object) => {
        return w.innerHeight;
      };
    } else {
      WindowHelper.innerHeightFunc = (w: Object) => {
        return w.document.documentElement.offsetHeight;
      };
    }
    if ('outerHeight' in window) {
      WindowHelper.outerHeightFunc = (w: Object) => {
        return w.outerHeight;
      };
    } else {
      WindowHelper.outerHeightFunc = WindowHelper.innerHeightFunc;
    }
    if ('clientWidth' in window) {
      WindowHelper.clientWidthFunc = (w: Object) => {
        return w['clientWidth'];
      };
    } else {
      WindowHelper.clientWidthFunc = (w: Object) => {
        return w.document.documentElement.clientWidth;
      };
    }
    if ('clientHeight' in window) {
      WindowHelper.clientHeightFunc = (w: Object) => {
        return w['clientHeight'];
      };
    } else {
      WindowHelper.clientHeightFunc = (w: Object) => {
        return w.document.documentElement.clientHeight;
      };
    }
    if (ss.isValue(window.self.pageXOffset)) {
      WindowHelper.pageXOffsetFunc = (w: Object) => {
        return w.pageXOffset;
      };
    } else {
      WindowHelper.pageXOffsetFunc = (w: Object) => {
        return w.document.documentElement.scrollLeft;
      };
    }
    if (ss.isValue(window.self.pageYOffset)) {
      WindowHelper.pageYOffsetFunc = (w: Object) => {
        return w.pageYOffset;
      };
    } else {
      WindowHelper.pageYOffsetFunc = (w: Object) => {
        return w.document.documentElement.scrollTop;
      };
    }
    if ('screenLeft' in window) {
      WindowHelper.screenLeftFunc = (w: Object) => {
        return (<any>w).screenLeft;
      };
    } else {
      WindowHelper.screenLeftFunc = (w: Object) => {
        return w.screenX;
      };
    }
    if ('screenTop' in window) {
      WindowHelper.screenTopFunc = (w: Object) => {
        return (<any>w).screenTop;
      };
    } else {
      WindowHelper.screenTopFunc = (w: Object) => {
        return w.screenY;
      };
    }
    {
      const DefaultRequestName: string = 'requestAnimationFrame';
      const DefaultCancelName: string = 'cancelAnimationFrame';
      let vendors: string[] = ['ms', 'moz', 'webkit', 'o'];
      let requestFuncName: string = null;
      let cancelFuncName: string = null;
      if (DefaultRequestName in window) {
        requestFuncName = DefaultRequestName;
      }
      if (DefaultCancelName in window) {
        cancelFuncName = DefaultCancelName;
      }
      for (let ii = 0; ii < vendors.length && (requestFuncName === null || cancelFuncName === null); ++ii) {
        let vendor: string = vendors[ii];
        let funcName: string = vendor + 'RequestAnimationFrame';
        if (requestFuncName === null && funcName in window) {
          requestFuncName = funcName;
        }
        if (cancelFuncName === null) {
          funcName = vendor + 'CancelAnimationFrame';
          if (funcName in window) {
            cancelFuncName = funcName;
          }
          funcName = vendor + 'CancelRequestAnimationFrame';
          if (funcName in window) {
            cancelFuncName = funcName;
          }
        }
      }
      if (requestFuncName !== null) {
        WindowHelper.requestAnimationFrameFunc = (callback: () => void) => {
          return window[requestFuncName](callback);
        };
      } else {
        WindowHelper.setDefaultRequestAnimationFrameImpl();
      }
      if (cancelFuncName !== null) {
        WindowHelper.cancelAnimationFrameFunc = (animationId: number) => {
          window[cancelFuncName](animationId);
        };
      } else {
        WindowHelper.cancelAnimationFrameFunc = (id: number) => {
          window.clearTimeout(id);
        };
      }
    }}

  public constructor(window: Object) {
    this.window = window;
  }

  /**
   * Gets the window pageXOffset, or equivalent.
   */
  public get pageXOffset(): number {
    return WindowHelper.pageXOffsetFunc(this.window);
  }

  /**
   * Gets the window pageYOffset, or equivalent.
   */
  public get pageYOffset(): number {
    return WindowHelper.pageYOffsetFunc(this.window);
  }

  /**
   * Gets the window pageXOffset, or equivalent.
   */
  public get clientWidth(): number {
    return WindowHelper.clientWidthFunc(this.window);
  }

  /**
   * Gets the window pageYOffset, or equivalent.
   */
  public get clientHeight(): number {
    return WindowHelper.clientHeightFunc(this.window);
  }

  /**
   * Gets the window innerWidth, or equivalent.
   */
  public get innerWidth(): number {
    return WindowHelper.innerWidthFunc(this.window);
  }

  /**
   * Gets the window outerWidth, or innerWidth if unavailable
   */
  public get outerWidth(): number {
    return WindowHelper.outerWidthFunc(this.window);
  }

  /**
   * Gets the window innerHeight, or equivalent.
   */
  public get innerHeight(): number {
    return WindowHelper.innerHeightFunc(this.window);
  }

  /**
   * Gets the window outerHeight, or innerHeight if unavailable.
   */
  public get outerHeight(): number {
    return WindowHelper.outerHeightFunc(this.window);
  }

  /**
   * Gets the window screenLeft, or equivalent.
   */
  public get screenLeft(): number {
    return WindowHelper.screenLeftFunc(this.window);
  }

  /**
   * Gets the window screenTop, or equivalent
   */
  public get screenTop(): number {
    return WindowHelper.screenTopFunc(this.window);
  }

  /**
   * Gets the window's self.
   * This method only exists to support testing.
   */
  public static get windowSelf(): Object {
    return window.self;
  }

  /**
   * Get the current window or document Selection object
   * @returns The selection object, or null if none
   */
  public static get selection(): Selection {
    if ((typeof (window['getSelection']) === 'function')) {
      return window.getSelection();
    } else
      if ((typeof (document['getSelection']) === 'function')) {
        return document.getSelection();
      }
    return null;
  }

  /**
   * Calls {@link Window.Close} on the given window.
   * This method only exists to support testing.
   */
  public static close(window: Object): void {
    window.close();
  }

  /**
   * Gets the window opener.
   * This method only exists to support testing.
   */
  public static getOpener(window: Object): Object {
    return window.opener;
  }

  /**
   * Gets the current window location.
   * This method only exists to support testing as the standard window.location cannot be mocked safely.
   */
  public static getLocation(window: Object): Location {
    return window.location;
  }

  /**
   * Gets the current window location.
   * This method only exists to support testing as the standard window.location cannot be mocked safely.
   */
  public static getPathAndSearch(window: Object): string {
    return window.location.pathname + window.location.search;
  }

  /**
   * Sets the href property on the window's location.
   * This method only exists to support testing as the standard window.location cannot be mocked safely.
   * @param window The window containing the location
   * @param href The href to set
   */
  public static setLocationHref(window: Object, href: string): void {
    window.location.href = href;
  }

  /**
   * Calls {@link Location.Replace} on the given window.
   * This method only exists to support testing as the standard window.location cannot be mocked safely.
   */
  public static locationReplace(window: Object, url: string): void {
    window.location.replace(url);
  }

  public static open(href: string, target: string): Object;

  /**
   * Calls {@link Window.Open}.
   * This method only exists to support testing as the standard window.location cannot be mocked safely.
   * @param href 
   * @param target 
   * @param options 
   */
  public static open(href: string, target: string, options?: string): Object {
    return window.open(href, target, options);
  }

  /**
   * Calls window.location.reload.
   * This method only exists to support testing as the standard window.location cannot be mocked safely.
   */
  public static reload(w: Object, forceGet: boolean = false): void {
    w.location.reload(forceGet);
  }

  /**
   * Requests an animation frame
   * Falls back to using Window.SetTimeout if the browser does not support RequestAnimationFrame
   * @param action Action to execute
   * @returns animation id used to cancel the animation
   */
  public static requestAnimationFrame(action: () => void): number {
    return WindowHelper.requestAnimationFrameFunc(action);
  }

  /**
   * Cancels an animation
   * @param animationId id of animation to cancel
   */
  public static cancelAnimationFrame(animationId: number): void {
    if (ss.isValue(animationId)) {
      WindowHelper.cancelAnimationFrameFunc(animationId);
    }
  }

  /**
   * This method exists so that Window.SetTimeout can be mocked
   */
  public static setTimeout(callback: () => void, milliseconds: number): void {
    window.setTimeout(callback, milliseconds);
  }

  public static addListener(windowParam: EventTarget, eventName: string, messageListener: HtmlEventHandler): void {
    if ('addEventListener' in windowParam) {
      windowParam.addEventListener(eventName, messageListener, false);
    } else {
      (<any>windowParam).attachEvent('on' + eventName, messageListener);
    }
  }

  public static removeListener(window: EventTarget, eventName: string, messageListener: HtmlEventHandler): void {
    if ('removeEventListener' in window) {
      window.removeEventListener(eventName, messageListener, false);
    } else {
      (<any>window).detachEvent('on' + eventName, messageListener);
    }
  }

  /**
   * Sets the RequestAnimationFrame implementation to the fallback if there is no browser support.
   * This is in its own method so the closure over 'lastTime' is smaller.
   */
  private static setDefaultRequestAnimationFrameImpl(): void {
    let lastTime: number = 0;
    WindowHelper.requestAnimationFrameFunc = (callback: () => void) => {
      let curTime: number = <number>(new Date().getTime());
      let timeToCall: number = Math.max(0, 16 - (curTime - lastTime));
      lastTime = curTime + timeToCall;
      let id: number = window.setTimeout(callback, timeToCall);
      return id;
    };
  }

  /**
   * Clears the selection object at the window/document level
   */
  public static clearSelection(): void {
    let selection: Selection = WindowHelper.selection;
    if (selection !== null) {
      if ((typeof (selection['removeAllRanges']) === 'function')) {
        selection.removeAllRanges();
      } else
        if ((typeof (selection['empty']) === 'function')) {
          selection['empty']();
        }
    }
  }
}

// Call the static constructor
WindowHelper.__ctor();
